### PR TITLE
Adds Script to Generate Component Docs

### LINKED ISSUE #592: Can we auto-document component API on the docs site?

### PR DESCRIPTION
**Pull request by redreceipt**
## DESCRIPTION

Kapture 2019-03-20 at 13 12 20

### What does this PR do, or why is it needed?

This will take JSDoc descriptions from our components and spit out markdown files with prop definitions that we can use on our doc site. It also includes a step in Travis that will automatically merge new docs into the site. Fully automated 🙌 

### How do I test this PR?

Add a JSDoc to one of our ui-kit components and then run `yarn generate-docs` from the root. Then check `packages/apollos-ui-kit/docs/generated/` for your new doc.

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [x] Closes #592
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [ ] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by conrad-vanl** [comment]
@redreceipt where does this put the docs, and how can we integrate this with the doc site?

**Comment by redreceipt** [comment]
It puts them in local directory, I wanted to discuss possible strategies
though. Maybe the docs repo could clone the ui-kit repo to do it automated?

**Comment by conrad-vanl** [comment]
I wonder if we can look at other projects we know that use docusaurus for insight, like react-native. Looks like the `react-native` has its own `gh-pages` branch...i bet they automate that somehow. I'm sure a simpler project on docusaurus would be easier to reverse engineer though.

**Comment by redreceipt** [comment]
Yea I agree. I'll do some research. Maybe it could be a githook or Travis could even do it for us or something. I'm assuming long term we'll probably just call the main repo `ApollosProject` or something and we could move the docs in there, is that what you're thinking?

**Comment by redreceipt** [comment]
React Native has a separate repo for the source files of their site https://github.com/facebook/react-native-website

**Comment by redreceipt** [comment]
@conrad-vanl what do you think about a flow along these lines. I would script this up in bash or something:
- pre-commit hook (using Husky) -> (`gen-docs.sh`)
- look through modified files and generate docs for them
- move them to `ui-kit/docs` dir
- user could submit PR to `ApollosProject/docs` or optionally we could automate that with `github/hub`

Still waiting on answers from a few different sources to see if there's anything smoother than that already out there.

**Comment by conrad-vanl** [comment]
@redreceipt did you get answers from any of those sources? Heh.

The only thing I _don't_ like about that list you set is it still involves a bit of manual intervention. I think the requirements I'd want to set are:

1. The process is fully automatic
2. Docs are tied to a specific version - docs really only need to be generated whenever we create a new version, and we get a version drop down on the doc site (like react-native)
2. The process is deterministic...same input === same output. In other words, it's okay if ALL of the component docs get re-generated every new version we cut, as long as the result is always the same. 

I'm happy to jump on a 📞 and brainstorm some ideas. Could do something like a `gh-pages` branch or `docs` folder on the `-prototype` repo that gets generated with each version, and then set that up as like a `submodule` on the `docs` repo. Probably some better options....

**Comment by redreceipt** [comment]
@conrad-vanl yea let's chat, the only piece of my suggested flow that isn't automated is the pull-request step into `ApollosProject/docs`, but it could be using the Github Pull-Request API or something. Or to fully automate it, we could just merge. That would all be handled by the bash (or python 😄) script. If we used a submodule, it would still be a manual process to pull the updated files. I'm not super concerned about items 2 and 3 in your requirements, should be good there.

And I spoke with one of the leads at Docusaurus, this use case seems like news to him but he liked the idea. Wants to build it into Docusaurus so that would be cool.

**Comment by conrad-vanl** [comment]
@redreceipt OKAY, I re-read your list above. I think I misread them the first time.

So here are my thoughts, then I think you can move forward on what you're proposing:

1. My experience with pre-commit hooks are they can be a pain for the developer. But perhaps this is something that could run on our CI layer by travis, and only applied on merge to master? So we merge a PR ==> script gets triggered on Travis as a part of our "build" phase, and that script can trigger a PR for the new docs over on the `docs` repo.

2. I'm also okay skipping the PR and going straight to merge

3. Could eventually go as far as adding a build step to our PR on `-prototype` that allows you to "preview" the generated docs from that PR. Kind of like the deploy previews on netlify

**Comment by redreceipt** [comment]
@conrad-vanl I think this should work. There are however a few things I don't love:

1. The `docs/buildDocs.sh` is a node script and is in dire need of refactoring. I am by no means a node expert and it was mostly a copy and paste.
2. This requires us to turn off all merge checks and PR approvals on `master` of the docs repo.
3. I'm using my GitHub creds in Travis to do the dirty work. While they are secure, I feel like it should be abstracted from me long term.

Thoughts???

**Comment by redreceipt** [comment]
[Generate Docs](https://trello.com/c/xyWPt1lf/5-generate-docs)

**Comment by conrad-vanl** [comment]
This is something I think @vinnyjth should be looking at since it pertains to build processes. Throwing him on it. He's supposed to be out this week which probably means this gets delayed a bit longer, but I think he's the right person to make the call here.

**Comment by redreceipt** [comment]
TODO: I still need to write the headers that the docusaurus content needs to display

**Comment by redreceipt** [comment]
@vinnyjth we *could* automate a PR if that's what you prefer by installing the [github/hub](https://github.com/github/hub) application on travis as part of the steps?

**Comment by vinnyjth** [comment]
1. I'll give it a quick review :) 
3. Not bothered by that. 
4. Would it be possible to, every time we generate docs, make a list of all generated docs and regenerate the `sidebar.js`?

File: packages/apollos-ui-kit/docs/buildDocs.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env node
+
+const fs = require('fs');
+const path = require('path');
+const generateMarkdown = require('./generateMarkdown');
+
+let json = '';
```
Comment (vinnyjth) [inline packages/apollos-ui-kit/docs/buildDocs.sh]: Small nitpick, but I would move these lines down to line `40` so all of the event handling is one place :)

File: packages/apollos-ui-kit/docs/buildDocs.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+process.stdin.setEncoding('utf8');
+process.stdin.on('readable', () => {
```
Comment (vinnyjth) [inline packages/apollos-ui-kit/docs/buildDocs.sh]: 👍 Really not a better way to do this without external libs.

File: packages/apollos-ui-kit/docs/buildDocs.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  let ext;
+  while ((ext = path.extname(name))) {
+    name = name.substring(0, name.length - ext.length);
+  }
+  return name;
+}
+
+function buildDocs(api) {
+  // api is an object keyed by filepath. We use the file name as component name.
+  for (const filepath in api) {
```
Comment (vinnyjth) [inline packages/apollos-ui-kit/docs/buildDocs.sh]: This is fine, but maybe a little less typical on our team. I tend to write this as `Object.keys(api).forEach((key) => ...)`
Some people would prefer what you wrote, and some people prefer what I normally write. I personally don't care and think what you have is fine :) 

Only reason it matters is if the object has inherited keys (from up the prototype chain), More info here: 
https://www.reddit.com/r/javascript/comments/8emf94/forin_vs_objectkeys/dxwecvs/

**Comment by redreceipt** [comment]
@vinnyjth ...yea we could generate `sidebars.json`. I'll just have to flex my tiny Node muscles.

What do you think about using [github/hub](https://github.com/github/hub) to automate a PR into docs instead? Necessary or overkill?

**Comment by vinnyjth** [comment]
@redreceipt I believe in you 💯 

In regard to Hub, I have no preference. Let's look to doing that in the future if we have a need for it. RIght now pushing straight to master sounds fine by me.

**Comment by redreceipt** [comment]
@vinnyjth ok, go through it with a fine toothed comb but I think we're good now!

File: packages/apollos-ui-kit/docs/generateMarkdown.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2015, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
```
Comment (redreceipt) [inline packages/apollos-ui-kit/docs/generateMarkdown.js]: @vinnyjth I did make a couple changes to this script, could we remove this license stuff from Facebook?

Comment (vinnyjth) [reply packages/apollos-ui-kit/docs/generateMarkdown.js]: Yah, you can go ahead and remove it ;)

**Comment by vinnyjth** [comment]
@redreceipt Dang! I just forgot about something that we need to add :/ The script in your travis.yml should only run in `master`. 

Here are my two ideas on how to solve: 

Can you move the stuff you have in the travis.yml into a `step` that only runs
Check if you're in `master`, and exit early if you're not.

**Comment by redreceipt** [comment]
@vinnyjth I actually think we're good here. It's my [understanding](https://docs.travis-ci.com/user/deployment/custom/) The `after_success` step only occurs after all stages have completed. And our `Deploy Api` stage only happens with a push to `master`. 

Screen Shot 2019-04-24 at 9 58 40 AM

Does that seem right to you? Plus I feel like if it wasn't protected, it would have ran the script with every push to this branch but it hasn't been.

**Comment by redreceipt** [comment]
@vinnyjth alright it's now a step, give it a once over

**Pull request merged**